File: Challenges/Python/GroupByAnagram.py
class Solution(object):
    def groupAnagrams(self, strs):
        results = dict()
        returnValue = []

        for string in strs:
            stringTokens = list(string)
            stringTokens.sort()
            orderedString = "".join(stringTokens)
            if results.__contains__(orderedString):
                entry = results.get(orderedString)
                entry.append(string)
                results[orderedString] = entry
            else:
                results[orderedString] = [string]

        for result in results.values():
            returnValue.append(result)

        return returnValue

File: Challenges/JavaScript/GroupByAnagram.js
var groupAnagrams = function(strs) {
   let results = new Map();
   let returnValue = [];

   strs.forEach(str => {
      let orderedTokens = str
         .split("")
         .sort()
         .join("");
      if (results.has(orderedTokens)) {
         let arr = results.get(orderedTokens);
         arr.push(str);
         results.set(orderedTokens, arr);
      } else results.set(orderedTokens, [str]);
   });

   results.forEach(element => {
      returnValue.push(element);
   });

   return returnValue;
};

const input = ["eat", "tea", "tan", "ate", "nat", "bat"];
const output = groupAnagrams(input);

console.log(output);
